src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs

```diff
@@ -874,11 +873,24 @@ public string ProcessName
                     else
                     {
                         // If we don't have a populated ProcessInfo, then get and cache the process name.
-                        _processName = ProcessManager.GetProcessName(_processId, _machineName);
+                        try
+                        {
+                            // Ensure that the process is not yet exited
+                            EnsureState(State.HaveNonExitedId);
+                            _processName = ProcessManager.GetProcessName(_processId, _machineName);
 
-                        if (_processName == null)
+                            // Ensure that a process name is available
+                            if (_processName == null)
+                            {
+                                throw new InvalidOperationException(SR.NoProcessInfo);
+                            }
+                        }
+                        catch
```

Comment:
If this case will be common, the previous change was actually a deoptimization, as it'll now pay the cost of both approaches plus an exception throw/catch. I'm more inclined to see the previous change reverted, but I'll leave that up to @jeffhandley and @adamsitnik.

Reply:
The new way usually works, is much more efficient (when it works),
and I believe also works sometimes when the old way does not, i.e. terminated processes.

I believe there is a race condition in the old way.
If you run a process that runs and exits quickly, you cannot
get its name. And .NET does not expose starting a process suspended (probably
because there is no way on Unix?) so there is no way to eliminate the race,
except the new way has no race.

Reply:
Of course, .NET could always start suspended and get the name, costing memory and again, it is really very inefficient, the old way gets information including name for almost all processes on the system, only to compare by pid and pick out the one it wanted. The new way only gets the information for the one specified process.

Reply:
For processes it start, .NET might also try to parse the name out of the command line.

Reply:
This approach looks like an ideal compromise. From what I can see, the actual work is only done once as the result is cached in in `_processName`.

Reply:
This approach looks like an ideal compromise. From what I can see, the actual work is only done once as the result is cached in in _processName.

Reply:
Only once if you get the name yes.
I believe that is/was further mitigation around cost _and_ failure modes in the old way.
i.e. if you managed to get the name, in a race, prior to termination, then you could have it after termination.
The caching seems less well motivated now, in the usual path, but not much harm..

We really should get Windows fixed here.
There is little point imho in restricting access via `QueryFullProcessImageNameW` when the data is available with no extra permissions via NtQuerySystemInformation.

But I realize .NET has to support older Windows so fixing Windows, if it happens, does not immediately suffice.